### PR TITLE
feat: OpenID Connect scheme

### PR DESCRIPTION
**Pull request by vanhoofmaarten**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

This PR adds support for authentication and authorization via the [OpenID Connect](https://openid.net/developers/specs/) specification by extending the existing OAuth2 scheme and adding 2 new main features:
1. Configuration document
    1. Fetch OpenID Connect server's configuration
    1. Set endpoints automatically
    1. Validate strategy configuration against fetched configuration
1. ID token support

### What is OpenID Connect?

> OpenID Connect 1.0 is a simple identity layer on top of the OAuth 2.0 protocol. It allows Clients to verify the identity of the End-User based on the authentication performed by an Authorization Server, as well as to obtain basic profile information about the End-User in an interoperable and REST-like manner. https://openid.net/connect/

#### 1. Configuration document.
See also: https://openid.net/specs/openid-connect-discovery-1_0.html#rfc.section.4

##### Fetch OpenID Connect server's configuration
On the initialization of a strategy with the `openIDConnect` scheme, Nuxt will fetch the OpenID Connect configuration document from the Authorization Server, declared by the `endpoints.configurationDocument` option and will save it in storage. In Nuxt Universal mode, the configuration document is fetched server-side and hydrated to the client.

##### Set endpoints automatically
Endpoints declared in the configuration document are appended to the strategy `endpoints` configuration. Endpoints declared in `nuxt.config` have a higher priority and overwrite endpoints from the configuration document.

##### Validation
The configuration document is validated against the following `nuxt-auth` configuration:

| Auth config | OpenID Connect Server config |
| --- | --- |
`responseType` | `response_type_supported`
`scope` | `scopes_supported`
`grantType` | `grant_types_supported`
`acrValues` | `acr_values_supported`

If not valid, `nuxt-auth` will throw warnings.

#### ID Token support
The [OpenID Connect RP-Initiated Logout 1.0](https://openid.net/specs/openid-connect-rpinitiated-1_0.html) specification, defines that a `id_token_hint` must be provided to the OpenID provider to also logout a user at the OpenID Provider. Although this specification is still a draft, it has already been implemented by widely used OpenID Connect Authentication frameworks like [MSFT Identity Server](https://identityserver4.readthedocs.io/en/latest/endpoints/endsession.html) and RedHat's [Keycloak](https://www.keycloak.org/docs-api/11.0/javadocs/index.html). To be able to send an ID-token, we first need to save one on authentication.

I've duplicated and extended the `Token` class with an additional method `userInfo`. Because we now can decode the information from an ID token and `nuxt-auth` check it's status on mount, it is also possible to set the user-info from the ID token without making an additional request to the OpenID Connect Server. See: https://github.com/vanhoofmaarten/auth-module/blob/241feec61fd6ca1c69b3ffbe0bb36cbe3ec1185b/src/schemes/openIDConnect.ts#L162

_PS. As I'm writing this PR, I think I see quite a bit of duplicated code. I might be able to abstract this by abstracting this to `AuthToken` and `IdToken` class which both extend a base class `Token`. Maybe something for later._

#### `oidc-provider`
To test this feature, I've added `oidc-provider`, a [OpenID Certified](https://openid.net/certification/#LogoutOPTable) OpenID Connect Provider with mocking capabilities to the demo and a new E2E test fixture.

**Demo**
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/demo/nuxt.config.ts#L11
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/demo/nuxt.config.ts#L152
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/demo/api/oidcmockserver.js

**Test fixture**
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/test/fixture/nuxt.config.js#L3
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/test/fixture/nuxt.config.js#L8
- https://github.com/vanhoofmaarten/auth-module/blob/feature/oidc/test/fixture/nuxt.config.js#L46

#### `jest-puppeteer`
To simplify the puppeteer integration into the E2E tests, I've added `jest-puppeteer` into the testing suite. This feature is tested in both Universal as SPA mode.

## Checklist:
- [x] My change requires a change (addition) to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

This PR resolves the feature request in #503.

Very much looking forward to your feedback.

### CODE REVIEW

**Comment by vanhoofmaarten** [comment]
Strange that the tests aren't passing. I've enabled action

**Comment by sjorsjes** [comment]
Working on implementing OIDC in a Nuxt application using openid-client, but I would love to see this PR merged!

**Comment by bmulholland** [comment]
@vanhoofmaarten Apologies for the slow response here. We've updated some things in the meantime and fixed CI - could you please fix the conflicts with the `dev` branch? After that CI should pass again. I haven't reviewed the code yet - will do that after conflicts are fixed and CI is passing.

**Comment by vanhoofmaarten** [comment]
Hi @bmulholland, 

I've managed to resolve the conflicts with the `dev` branch and improved the test, so they pass on CI as well.

Awaiting your review.

**Comment by WhiteRabbitDE** [comment]
Is this PR alive ? That would be an awesome addition !

**Comment by bmulholland** [comment]
Alive, yes, but I need to put some solid time into it and just haven't had that. Has anyone tried out this branch and can confirm it works? That would be a good start.

**Comment by vanhoofmaarten** [comment]
I'll better fix those conflicts first. 😅 I'll try to dedicate some time this week to this PR.

**Comment by bmulholland** [comment]
Thanks for fixing them several times @vanhoofmaarten, I appreciate it. I commit to reviewing it promptly once you have.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -51,11 +51,8 @@ jobs:
       - name: Build Module
         run: yarn build
 
-      - name: Build Fixture
-        run: yarn nuxt build test/fixture
-
       - name: Tests
-        run: yarn jest
+        run: yarn jest --runInBand
```
Comment (bmulholland) [inline .github/workflows/ci.yml:55]: Was there anything in this PR causing issues in CI? We'd want to keep the CI workflow the same.

File: demo/api/oidcmockserver.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* eslint-disable no-console */
```
Comment (bmulholland) [inline demo/api/oidcmockserver.js:1]: Can we just disable console for the specific lines it's necessary?

File: demo/api/oidcmockserver.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+const { Provider } = require('oidc-provider')
+const defu = require('defu')
+
+// Suppress oidc-provider errors on test run
```
Comment (bmulholland) [inline demo/api/oidcmockserver.js:5]: What errors need to be hidden? Won't this hide legitimate issues, too?

File: demo/nuxt.config.ts [hunk 4]
```diff
@@ -129,6 +144,18 @@ export default {
         responseType: 'code',
         grantType: 'authorization_code',
         clientId: 'test-client'
+      },
+      oidcmock: {
+        scheme: 'openIDConnect',
+        responseType: 'code',
+        scope: ['openid', 'profile', 'offline_access'],
```
Comment (bmulholland) [inline demo/nuxt.config.ts:151]: Some of these options, like scope and response type, seem like good default OIDC options. What about setting those as defaults and removing them from the necessary config? Makes it a little easier to use.

File: jest-puppeteer.config.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  browserContext: 'incognito'
```
Comment (bmulholland) [inline jest-puppeteer.config.js:2]: Why?

File: jest.config.js [hunk 6]
```diff
@@ -1,13 +1,14 @@
+const { jsWithBabel: tsjPreset } = require('ts-jest/presets')
+
 module.exports = {
-  preset: 'ts-jest',
+  preset: 'jest-puppeteer',
```
Comment (bmulholland) [inline jest.config.js:4]: Why switch?

File: package.json [hunk 7]
```diff
@@ -31,7 +31,7 @@
     "lint-fix": "yarn lint --fix",
     "release": "yarn clean && yarn build && standard-version && git push --follow-tags && npm publish",
     "start": "nuxt-ts start demo",
-    "test": "yarn lint && yarn build && nuxt build test/fixture && jest"
+    "test": "yarn lint && yarn build && jest --runInBand"
```
Comment (bmulholland) [inline package.json:34]: Similar question about if we can keep CI the same.

File: package.json [hunk 8]
```diff
@@ -65,9 +67,11 @@
     "express-jwt": "latest",
     "get-port": "latest",
     "jest": "latest",
+    "jest-puppeteer": "latest",
     "jsdom": "latest",
     "lodash.get": "latest",
     "nuxt-edge": "latest",
+    "oidc-provider": "^6.29.9",
```
Comment (bmulholland) [inline package.json:74]: Why not "latest"?

File: src/inc/configuration-document.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+import Storage from '../core/storage'
+
+class ConfigurationDocumentRequestError extends Error {
+  constructor() {
+    super('Error loading OpenIDConnect configuration document')
+    this.name = 'ConfigurationDocumentRequestError'
+  }
+}
+
+// eslint-disable-next-line no-console
```
Comment (bmulholland) [inline src/inc/configuration-document.ts:12]: I think only the disable on line 14 is necessary, is that right?

File: src/inc/configuration-document.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+const ConfigurationDocumentWarning = (message: string) =>
+  // eslint-disable-next-line no-console
+  console.warn(`[AUTH] [OPENID CONNECT] Invalid configuration. ${message}`)
```
Comment (bmulholland) [inline src/inc/configuration-document.ts:15]: Thanks for providing this warning! Nice to have error checking like this.

File: src/inc/configuration-document.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  public key: string
+
+  constructor(scheme: Scheme, storage: Storage) {
+    this.scheme = scheme
+    this.$storage = storage
+    this.key = '_configuration_document.' + this.scheme.name
+  }
+
+  _set(value: OpenIDConnectConfigurationDocument | boolean) {
+    return this.$storage.setState(this.key, value)
```
Comment (bmulholland) [inline src/inc/configuration-document.ts:29]: @JoaoPedroAS51 Is this consistent with what we do elsewhere?

File: src/inc/configuration-document.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    this.name = 'ConfigurationDocumentRequestError'
+  }
+}
+
+// eslint-disable-next-line no-console
+const ConfigurationDocumentWarning = (message: string) =>
+  // eslint-disable-next-line no-console
+  console.warn(`[AUTH] [OPENID CONNECT] Invalid configuration. ${message}`)
+
+export default class ConfigurationDocument {
```
Comment (bmulholland) [inline src/inc/configuration-document.ts:17]: Could you write a short comment explaining what configuration doc is, what it does, and why it's necessary?

**Review by bmulholland — CHANGES_REQUESTED** [review]
I'm only partway through my review, but have it leave it here for now. Posting the comments I have so far to keep things moving.

File: src/inc/configuration-document.ts (hunk 11 above)
Comment (JoaoPedroAS51) [reply src/inc/configuration-document.ts:29]: We should not use `setState` directly. Use `setUniversal` instead.

File: src/inc/configuration-document.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    this.$storage = storage
+    this.key = '_configuration_document.' + this.scheme.name
+  }
+
+  _set(value: OpenIDConnectConfigurationDocument | boolean) {
+    return this.$storage.setState(this.key, value)
+  }
+
+  get(): OpenIDConnectConfigurationDocument {
+    return this.$storage.getState(this.key)
```
Comment (JoaoPedroAS51) [inline src/inc/configuration-document.ts:33]: Same here. Use `getUniversal` instead.

File: src/inc/configuration-document.ts [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  set(value: OpenIDConnectConfigurationDocument | boolean) {
+    this._set(value)
```
Comment (JoaoPedroAS51) [inline src/inc/configuration-document.ts:37]: I think this could be simplified. We don't need to use a private method here.

```ts
set(value: OpenIDConnectConfigurationDocument | boolean) {
  return this.$storage.setUniversal(this.key, value)
}
```

File: src/inc/configuration-document.ts [hunk 15]
```diff
@@ -0,0 +1,116 @@
+import defu from 'defu'
+import type { Scheme, OpenIDConnectConfigurationDocument } from '../index'
+import Storage from '../core/storage'
+
+class ConfigurationDocumentRequestError extends Error {
```
Comment (JoaoPedroAS51) [inline src/inc/configuration-document.ts:5]: To maintain consistency, can you move this class to a separate file?

File: src/schemes/openIDConnect.ts [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+  idToken: {
+    property: 'id_token',
+    maxAge: 60 * 60 * 24 * 30,
+    prefix: '_id_token.',
+    expirationPrefix: '_id_token_expiration.'
+  },
+  codeChallengeMethod: 'S256'
+}
+
+export default class OpenIDConnectScheme extends Oauth2Scheme<typeof DEFAULTS> {
```
Comment (JoaoPedroAS51) [inline src/schemes/openIDConnect.ts:24]: There is a new way to extend scheme options. Take a look at https://github.com/nuxt-community/auth-module/blob/dev/src/schemes/oauth2.ts#L39

**Review by JoaoPedroAS51 — CHANGES_REQUESTED** [review]
Hi @vanhoofmaarten ! Nice work! :smiley: 

I added some comments. I see that there's already things to do, so I'll do a better review later. Don't want to overwhelm you haha

File: src/inc/configuration-document.ts (hunk 13 above)
Comment (vanhoofmaarten) [reply src/inc/configuration-document.ts:33]: I've implemented `setState` in favor of `setUniversal` because I didn't want to save the configuration document, which can contain a fairly large amount of data for a cookie, in a cookie. This prevents that requests headers will become to large.

The configuration-document of the demo for example:
https://gist.github.com/vanhoofmaarten/fc1ecb5bb59a7bcd9364e6233fffb8df

In universal mode, I chose to fetch and set the configuration document state on server and use the `beforeNuxtRender` method to prepare the document for hydration to the client.

How shall I continue?

File: src/inc/configuration-document.ts (hunk 14 above)
Comment (vanhoofmaarten) [reply src/inc/configuration-document.ts:37]: Agreed.

File: src/schemes/openIDConnect.ts (hunk 16 above)
Comment (vanhoofmaarten) [reply src/schemes/openIDConnect.ts:24]: Thanks for the offer. I think I managed the upgrade path.

File: demo/api/oidcmockserver.js (hunk 3 above)
Comment (vanhoofmaarten) [reply demo/api/oidcmockserver.js:5]: Agreed.

File: jest-puppeteer.config.js (hunk 5 above)
Comment (vanhoofmaarten) [reply jest-puppeteer.config.js:2]: This will be removed, as I'll revert to the newly introduced `playwright` integration.

File: demo/api/oidcmockserver.js (hunk 3 above)
Comment (vanhoofmaarten) [reply demo/api/oidcmockserver.js:5]: But! When running tests, even when all tests pass, the console is cluttered with the following messages. I'd like to suppress these, because it's fairly obvious we're using the development mode.

```
console.warn
      WARNING: a quick start development-only in-memory adapter is used, you MUST change it in order to not lose all stateful provider data upon restart and to be able to share these between processes

      16 |   const redirectUri = `${appBaseUrl}${redirect.callback}`
      17 |
    > 18 |   return new Provider(baseUrl, {
         |          ^
      19 |     routes: {
      20 |       authorization: '/connect/authorize',
      21 |       token: '/connect/token',

      at Object.warn (node_modules/oidc-provider/lib/helpers/attention.js:24:11)
      at Provider.initializeAdapter (node_modules/oidc-provider/lib/helpers/initialize_adapter.js:8:15)
      at new Provider (node_modules/oidc-provider/lib/provider.js:128:23)
      at provider (demo/api/oidcmockserver.js:18:10)
      at oidcMockServer (demo/api/oidcmockserver.js:47:45)
      at handler (test/fixture/nuxt.config.js:8:45)
      at call (../test-utils/node_modules/connect/index.js:239:7)
      at next (../test-utils/node_modules/connect/index.js:183:5)

    console.warn
      WARNING: a quick start development-only signing keys are used, you are expected to provide your own in configuration "jwks" property

      16 |   const redirectUri = `${appBaseUrl}${redirect.callback}`
      17 |
    > 18 |   return new Provider(baseUrl, {
         |          ^
      19 |     routes: {
      20 |       authorization: '/connect/authorize',
      21 |       token: '/connect/token',

      at Object.warn (node_modules/oidc-provider/lib/helpers/attention.js:24:11)
      at Provider.initializeKeystore (node_modules/oidc-provider/lib/helpers/initialize_keystore.js:53:15)
      at new Provider (node_modules/oidc-provider/lib/provider.js:129:24)
      at provider (demo/api/oidcmockserver.js:18:10)
      at oidcMockServer (demo/api/oidcmockserver.js:47:45)
      at handler (test/fixture/nuxt.config.js:8:45)
      at call (../test-utils/node_modules/connect/index.js:239:7)
      at next (../test-utils/node_modules/connect/index.js:183:5)
```

File: package.json (hunk 8 above)
Comment (vanhoofmaarten) [reply package.json:74]: I'm finding several other versioned packages. How should I move forward with this newly introduced dependency?

File: src/inc/configuration-document.ts (hunk 11 above)
Comment (vanhoofmaarten) [reply src/inc/configuration-document.ts:29]: I've replied in your review below.

**Comment by vanhoofmaarten** [comment]
I've managed to merge the new changes on `dev` in to the PR, resolve or provide feedback most of the reviews.
Tests won't pass on CI, because this PR needs the following `test-utils` fix/feature  https://github.com/nuxt/test-utils/pull/84

**Comment by vanhoofmaarten** [comment]
Hi, quick update. Still working on the tests. Opened another [new PR on test-utils repo](https://github.com/nuxt/test-utils/pull/87) for the use-cases I need to cover. Slowly getting there.

**Comment by jpsc** [comment]
@vanhoofmaarten I would like to help test this. Is there a way to test this locally but not on the oidc mock server?

I've tried to add this to deps:

```json package.json
"@nuxtjs/auth-next": "vanhoofmaarten/auth-module",
```

But I get the error: `Please verify that the package.json has a valid "main" entry`

**Comment by vanhoofmaarten** [comment]
Hi @jpsc, I wasn't aware you could do that with NPM. Well, TIL then. If you `npm install` via that route, it won't install the `dist` folder that contains the compiled project as it is ignored by git.

I would suggest cloning the repo, building it and `npm link` it to your particular Nuxt-project.

PS. Thanks for the help!

**Comment by jpsc** [comment]
Currently running `npm run build` (tsc) will return the following error.

```shell 
src/schemes/oauth2.ts:51:10 - error TS2610: 'name' is defined as an accessor in class 'BaseScheme<{ name: string; accessType: any; redirectUri: any; logoutRedirectUri: any; clientId: any; audience: any; grantType: any; responseMode: any; acrValues: any; autoLogout: boolean; endpoints: { logout: string; authorization: string; token: string; userInfo: string; }; ... 4 more ...; codeChallengeMethod: stri...', but is overridden here in 'Oauth2Scheme' as an instance property.

51   public name
```

File: docs/content/en/schemes/openIDConnect.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+#### `property`
+
+- Default: `refresh_token`
+
+`property` can be used to specify which field of the response JSON to be used for value. It can be `false` to directly use API response or being more complicated like `auth.refresh_token`.
+
+#### `maxAge`
+
+- Default: `60 * 60 * 24 * 30`
```
Comment (jpsc) [inline docs/content/en/schemes/openIDConnect.md:146]: Does it make sense to align this with recent change on line 33 `src/schemes/openIDConnect.ts`

Comment (vanhoofmaarten) [reply docs/content/en/schemes/openIDConnect.md:146]: This is the same default `maxAge` as the OAuth2 scheme.

**Comment by vanhoofmaarten** [comment]
Hi @bmulholland and @JoaoPedroAS51. I've provided the requested changes and have some open questions. Could you take a look at these? I believe this PR is close to finished.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (bmulholland) [reply .github/workflows/ci.yml:61]: This comment was marked resolved, but the change is still there and the comment is unanswered?

File: package.json (hunk 8 above)
Comment (bmulholland) [reply package.json:81]: Yeah looks like we've been inconsistent here. Good for now.

File: src/schemes/oauth2.ts [hunk 18]
```diff
@@ -75,10 +75,10 @@ const DEFAULTS: SchemePartialOptions<Oauth2SchemeOptions> = {
   acrValues: null,
   autoLogout: false,
   endpoints: {
-    logout: '',
-    authorization: '',
-    token: '',
-    userInfo: ''
+    logout: null,
```
Comment (bmulholland) [inline src/schemes/oauth2.ts:78]: Why these changes?

File: package.json [hunk 19]
```diff
@@ -34,7 +34,7 @@
     "lint-fix": "yarn lint --fix",
     "release": "yarn build && standard-version && git push --follow-tags && npm publish",
     "start": "nuxt-ts start demo",
-    "test": "yarn lint && jest"
+    "test": "yarn lint && jest --runInBand"
```
[the remark this answers is not in the document]
Comment (vanhoofmaarten) [reply package.json:37]: I tried to run the tests in parallel mode (no `--runInBand`), but they didn't pass. Running them serially does the trick.

File: src/schemes/oauth2.ts (hunk 18 above)
Comment (vanhoofmaarten) [reply src/schemes/oauth2.ts:78]: Setting an empty string as default value will break the automatic setting of the endpoints by the configuration document.

https://github.com/nuxt-community/auth-module/blob/c177cf47a573b931d2f083e75ea1ea2a07e7e07d/src/inc/configuration-document.ts#L135-L140

Manually specified endpoints have priority over the automatic set endpoints. `defu` library sees an empty string as a set value and will override the automatically set value, hence the change to `null` as default value.

**Comment by vanhoofmaarten** [comment]
Hi @bmulholland  & @JoaoPedroAS51. I've provided your questions with an answer. 

I've had the opportunity of integrating custom versions of the v4 version of this library with both RedHat SSO\ KeyCloak and  .NET Core Identity Server OpenIDConnect providers of two major projects with an average user base of thousands of monthly users. This integration included creating support for PKCE, refresh-token in the previous iteration on both web as hybrid platforms (Cordova). I can confidently say, I've spent quite some time in the oath2 scheme code.

I haven't had the opportunity yet to integrate this new piece of code into a project. However, I provided several new E2E tests that test this functionality against a certified OpenIDConnect Provider, `oidc-provider` to ensure that everything works and will keep on working in the future.

File: package.json (hunk 19 above)
[the remark this answers is not in the document]
Comment (bmulholland) [reply package.json:37]: My experience with test suites is that issues like that can be a red flag, and they can compound in really difficult to diagnose ways. This change in particular will permanently increase the time to run the test suite. Why does it work when run in band? Can we fix the root cause?

**Comment by bmulholland** [comment]
Thanks for the additional info! I had a conversation with Joao and your comments addressed his concerns - thanks for that, too :)

Remaining steps here:

* Address my concern about the test suite, above
* My team will deploy this branch on our product; we'll merge this here once it's deployed to our production and confirmed to be working well

**Comment by JoaoPedroAS51** [comment]
Hi @vanhoofmaarten! Thank you for all your work on this :)

Just waiting for the remaining steps as pointed out by @bmulholland

**Comment by vanhoofmaarten** [comment]
@bmulholland I changed the tests from running serially back to parallel(y?). Locally all tests pass. Can you approve me for running the workflow on this PR, so we can run the CI tests? 

Did you have the opportunity yet of testing this PR with your app? I'm curious if there are any findings on your side.

**Review by bmulholland — APPROVED** [review]

**Comment by bmulholland** [comment]
Awesome - thank you! I don't see the usual request to approve CI, and the tests seem to have run (and passed) - am I missing something?

My team has had other priorities and hasn't gotten to trying this out; I'll do it myself today and report back.

**Comment by bmulholland** [comment]
@JoaoPedroAS51 Aha! Thank you :)

**Review by bmulholland — CHANGES_REQUESTED** [review]
Alright, I tested it out and it's working well - thanks again, this is a good contribution :)

Two minor fixes needed:
* Two updates to the docs, as commented
* When logging out, a flash of the redirected application logged out state (redirect) is shown _before_ being taken to the OpenID sign out page. This is a confusing user experience: the logged out state/redirect should only show to the user after they finish the logout via OpenID.

**Comment by marijngiesen** [comment]
Is there anything I can do to help speed this up?

**Comment by bmulholland** [comment]
@marijngiesen outstanding issues are as flagged above: two small updates to the docs files, and one larger investigation about the logout behaviour. Feel free to work on any of those.

**Comment by vanhoofmaarten** [comment]
Hey @marijngiesen, thank you for your offer. I'm having difficulties freeing up dedicated time to complete this PR.

> When logging out, a flash of the redirected application logged out state (redirect) is shown before being taken to the OpenID sign out page. This is a confusing user experience: the logged out state/redirect should only show to the user after they finish the logout via OpenID.

I've come across a similar use case in a project where we use the v4 of this module. There we did the following:

1. Fire logout method
2. send user to Provider for OpenID Connect logout
3. User gets redirected to additional `callbackLogoutUrl`
4. [`handleCallback`](https://github.com/vanhoofmaarten/auth-module/blob/7f3d56c92b841f1210fa6cb07dfe6e10ea36dcdb/src/schemes/openIDConnect.ts#L202) checks if the callBackUrl is for login or logout
5. If callbackUrl is for logout -> fire [`reset`](https://github.com/vanhoofmaarten/auth-module/blob/7f3d56c92b841f1210fa6cb07dfe6e10ea36dcdb/src/schemes/openIDConnect.ts#L158) method

This is on the presumption that the OpenIDConnect Provider has a `logoutRedirectURI` and correctly redirects the user on both logout and cancellation of the logout.

IMHO: To have a complete implementation, both logout flows, with and without the presence of a `logoutRedirectURI` should be present.

@marijngiesen @bmulholland What are your thoughts on this?

PS Shouldn't this behaviour also be present in the `OAuth2` scheme, as it is the same code / flow?

**Comment by bmulholland** [comment]
@vanhoofmaarten Thanks for all your work here, and for writing up your thoughts. We appreciate whatever you can find time for :)

WRT your comments about the logout redirect - I think I'm maybe missing some information.  How does what you're saying connect to the issue I flagged? Is the flow you mentioned a workaround? Or is it a fix and I just don't know the root cause of the bug? Can you help me connect the dots?

**Comment by bmulholland** [comment]
(Apparently if you accidentally hold shift while submitting the form with a keyboard shortcut, it closes a PR!)

**Comment by miguel-a-calles-mba** [comment]
@vanhoofmaarten, is there any chance we can get this moving? I am looking forward to using your OIDC scheme.

**Comment by bmulholland** [comment]
@miguel-a-calles-mba The necessary work is all outlined above, and your contributions to help resolve it would certainly be welcomed.

**Comment by miguel-a-calles-mba** [comment]
@bmulholland, how do I contribute? I tried to make changes yesterday but I do not seem to have write privileges for this PR. Please advise.

**Comment by bmulholland** [comment]
Thanks for offering! I don't think I can fix your permissions for that, but feel free to create a fork of the branch in your own github and post a link here.

**Comment by miguel-a-calles-mba** [comment]
@bmulholland, I hope this is setup correctly. https://github.com/miguel-a-calles-mba/auth-module/tree/feature/oidc

File: src/schemes/oauth2.ts (hunk 18 above)
Comment (miguel-a-calles-mba) [reply src/schemes/oauth2.ts:78]: @bmulholland, is this comment still open?

**Comment by bmulholland** [comment]
Thanks for your thoughts. I was reflecting on this myself and agree that it's good enough for now, and we can make tweaks to that behaviour later. @vanhoofmaarten did you pull in all of @miguel-a-calles-mba's work? If so, from your perspective, is this PR ready to merge?

**Comment by vanhoofmaarten** [comment]
@bmulholland, I very much did pull in @miguel-a-calles-mba work. It's good to go for me!

**Comment by bmulholland** [comment]
Merged - thank you all for the help and changes along the way.

Note that deploys to npm are currently broken; you can try it out on the master branch until we get that fixed.

**Comment by AllanOricil** [comment]
@vanhoofmaarten I just tried to use it and I got this error

image


What could it be? 

I have the latest version

image

**Comment by bmulholland** [comment]
@AllanOricil 

> Note that deploys to npm are currently broken; you can try it out on the master branch until we get that fixed.

**Comment by AllanOricil** [comment]
@bmulholland  but the master branch is 4.9.1. You mean dev branch, right?

**Comment by bmulholland** [comment]
Yes, apologies 🙈

**Comment by AllanOricil** [comment]
@bmulholland I did this 

image

image

and got this error

image

What Im missing?

**Comment by AllanOricil** [comment]
ah ok, I have to clone the repo, checkout this branch, build it, then create a link on it.  Now Im sad :(

**Comment by AllanOricil** [comment]
@bmulholland I can't build it. Could you guys please add the build folder in the dev branch. This way nobody will need to go through this

image

**Comment by tnabil** [comment]
@bmulholland Thanks for the great work, guys! Any chance this is making it into an npm release soon?

**Comment by muhmuhhum** [comment]
@bmulholland any updates regarding the npm packages ?

**Comment by bmulholland** [comment]
It's out :)
